Fix missing element types and shapes when autograd.Function has multiple tensor outputs

wschin opened the pull request:
When generating IR for autograd.Function, if the function has multiple outputs, a TupleUnpack may be inserted after the original function node, and Pytorch only assigns proper information (tensor element type and shape) to the TupleUnpack and forgets the original function node. In contrast, if autograd.Function only produces one output, the original function node may have tensor 
element type and shape in its output schema.

Before this PR:
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp -> output (tensor, dtype=float32, shape=[4, 5])
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp -> output_0 **(tensor)**, output_1 **(tensor)** -> TupleUnpack output_2 (tensor, dtype=float32, shape=[4, 5]), output_3 (tensor, dtype=float32, shape=[6, 7])

After this PR:
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp -> output (tensor, dtype=float32, shape=[4, 5])
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp ->output_0 **(tensor, dtype=float32, shape=[4, 5])**, output_1 **(tensor, dtype=float32, shape=[6, 7])** -> TupleUnpack output_2 (tensor, dtype=float32, shape=[4, 5]), output_3 (tensor, dtype=float32, shape=[6, 7])

wschin committed:
Fix missing element types and shapes when autograd.Function has multiple tensor outputs.
Note that all outputs are always tensors.

albanD commented:
cc @eellison do you know who could review this?

eellison commented:
cc @gmagogsfm

BowenBao commented:
cc @gmagogsfm @SplitInfinity friendly ping for review 

Usually for ONNX related updates we review and merge into onnx_ms_1 branch, but since this one interacts with autograd function and how it's traced, I think it's best for you to take a look.

BowenBao commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -514,6 +517,18 @@ static void _trace_post_record(
       }
     }
   }
+  // If TuplUunpack operator is created, we copy its output type back
```
```suggestion
  // If TupleUnpack operator is created, we copy its output type back
```

wschin committed:
Add a test and fix a typo

Make sure test runs

BowenBao commented:
@gmagogsfm ping ...

wschin committed:
Format

wschin commented on test/jit/test_tracer.py:
```diff
@@ -1407,6 +1406,36 @@ def forward(self, x):
+    def test_trace_multi_output_function(self):
+        # An autograd.Function with two outputs.
+        # It swaps inputs so we can check if shape
+        # handling is correct in TorchScript .
```
```suggestion
        # handling is correct in TorchScript.
```

wschin committed:
Update test/jit/test_tracer.py

gmagogsfm commented on test/jit/test_tracer.py:
```diff
@@ -1407,6 +1406,36 @@ def forward(self, x):
 
         traced = torch.jit.trace(Wrapper(), (torch.rand(3, 4),))
 
+    def test_trace_multi_output_function(self):
```
Does this test fail without shape information on original call note?

wschin replied:
It doesn't but it still checks the newly added part. If new change is wrong (I tried randomly assigned some shapes), this test may fail.

gmagogsfm replied:
In that case, could you use `FileCheck()` to ensure correct tensor information is in the IR? That would make the test fail in case of tensor properties being propagated differently

wschin replied:
I have added new logics to directly examine my change:
```
        for n in traced.graph.nodes():
            if n.kind() != 'prim::PythonOp':
                continue
            for o in n.outputs():
                if o.type().kind() != 'TupleType':
                    continue
                for e in o.type().elements():
                    # In this test, tensor elements in a tuple output 
                    # can't have unknown shapes.
                    assert e.dim()
```
This `assert` throws without my change. Basically, I check the IR before dumping it to file, so we don't need `FileCheck()` here.

pengwa replied:
Hi @gmagogsfm Could you please help review this change? thanks a lot.

gmagogsfm replied:
`FileCheck()` doesn't require dumping IR to a file actually, you can pass a string to it, it is generally nicer to write and read than logics that directly check for a pattern.

Here is an exmaple:
https://github.com/pytorch/pytorch/blob/22e3cc21e535e29deea091adf93e0408139b6aa3/test/test_jit.py#L496

wschin replied:
Done using `FileCheck`. Thanks.

wschin committed:
Modify test.
In addition to test if the created IR is runnable, we also
want to ensure "missing shapes are added back".

wschin committed:
Merge branch 'fix-type' of github.com:wschin/pytorch into fix-type

wschin committed:
Merge remote-tracking branch 'origin/master' into fix-type

wschin committed:
Merge remote-tracking branch 'origin/master' into fix-type

gmagogsfm commented on test/jit/test_tracer.py:
```diff
@@ -1407,6 +1406,47 @@ def forward(self, x):
+                for e in o.type().elements():
+                    # In this test, tensor elements in a tuple output 
+                    # can't have unknown shapes.
+                    assert e.dim()
```
Nit:
Please use `self.assert***` methods instead of raw `assert`

wschin replied:
I am trying to replace those checks with `FileCheck`, so these `assert`s will be removed. Thanks.

gmagogsfm reviewed (approved):
I think this is functionally sound, but there are a few nits on how the tests are written.

Sorry for the delay, message was buried in floods of GH emails and this PR dropped off from my GH review dashboard since last round of comment.

wschin committed:
Address comments

Fix

wschin commented:
@gmagogsfm, Thanks a lot for reviewing again. All comments are addressed now. Once the build passes, could you please merge this PR? Please let me know if concerns.

wschin committed:
Fix typo
